File: 0992. Subarrays with K Different Integers.java
// Hard
// Sliding Window
// Time:O(n), Space:O(k)
// https://leetcode.cn/problems/subarrays-with-k-different-integers/

import java.util.*;

class Solution {
    public int subarraysWithKDistinct(int[] nums, int k) {
        // 刚好k个不同整数 = 最多k个不同整数-最多k-1个不同整数
        // 可以用map或计数数组做
        return helper(nums, k) - helper(nums, k - 1);
    }

    private int helper(int[] nums, int k) {
        int n = nums.length;
        Map<Integer, Integer> map = new HashMap<>();
        int left = 0, res = 0;
        for (int right = 0; right < n; right++) {
            map.put(nums[right], map.getOrDefault(nums[right], 0) + 1);

            while (map.size() > k) {
                map.put(nums[left], map.get(nums[left]) - 1);
                if (map.get(nums[left]) == 0) {
                    map.remove(nums[left]);
                }
                left++;
            }
            res += right - left + 1;
        }
        return res;
    }
}

File: 0501-1000/0992. Subarrays with K Different Integers.py
# Hard
# Sliding Window. Hash Table
# Time:O(n), Space:O(n) 
# https://leetcode.cn/problems/subarrays-with-k-different-integers/

# exactly(k) = atMost(k) - atMost(k-1)

from typing import *

class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        def atMostK(k):
            from collections import defaultdict
            cnt = defaultdict(int)
            left = 0
            res = 0

            for right in range(len(nums)):
                if cnt[nums[right]] == 0: 
                    k -= 1
                cnt[nums[right]] += 1

                while k < 0:
                    cnt[nums[left]] -= 1
                    if cnt[nums[left]] == 0:
                        k += 1
                    left += 1
                
                res += right - left + 1
            return res
        
        return atMostK(k) - atMostK(k - 1)
